staging/src/k8s.io/api/resource/v1alpha2/types.go

```diff
@@ -462,3 +523,208 @@ type ResourceClaimTemplateList struct {
 	// Items is the list of resource claim templates.
 	Items []ResourceClaimTemplate `json:"items" protobuf:"bytes,2,rep,name=items"`
 }
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +k8s:prerelease-lifecycle-gen:introduced=1.30
+
+// NodeResourceSlice provides information about available
+// resources on individual nodes.
+type NodeResourceSlice struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard object metadata
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// NodeName identifies the node where the capacity is available.
+	// A field selector can be used to list only NodeResourceSlice
+	// objects with a certain node name.
+	NodeName string `json:"nodeName" protobuf:"bytes,2,name=nodeName"`
```

Comment:
Network-attached resources need further thought. There are several opens:
- What if such a resource is needed by "the cluster" instead of a pod? Then the scheduler is not involved and we have nothing that handles the allocation.
- Allocation and initialization of a network-attached resource might need to do vendor-specific work at the cluster level. Which component should do that? kubelet might not be the right place, even when assuming that the resource is used on a node.
- Because the resource is not owned by a node, we cannot use kubelet for publishing resource information.

I don't have answers for this, but my gut feeling is that NodeResourceSlices will continue to be used for node-local resources, with the special support that we are building into the apiserver to limit access (node authorizer, admission).

Reply:
I started thinking about how to handle network-attached resources and perhaps `ResourceSlice` (sic!) would also be usable for those.

The semantic of `NodeName` changes slightly: it becomes an optional field and will be required only for `ResourceSlices` accessed by the kubelet.

Let me add one commit at the end where I renamed the type. If we don't like that, I can rip it out again.